### PR TITLE
Add erase and force flash as flashing options

### LINKED ISSUE #513: Add --erase flag support

### PR DESCRIPTION
**Pull request by justinlampley**
image

Closes #513

### CODE REVIEW

**Review by pkendall64 — COMMENTED** [review]
Nice one 👍

**Comment by jurgelenas** [comment]
> image
> 
> Closes #513

What do you think about moving these checkboxes to the right of "Build & Flash" while making them smaller?

File: src/ui/views/ConfiguratorView/index.tsx [hunk 1]
```diff
@@ -423,6 +429,26 @@ const ConfiguratorView: FunctionComponent<ConfiguratorViewProps> = (props) => {
...
+  const eraseSupported = useMemo(() => {
+    if (!deviceTarget || !device) {
+      return false;
+    }
+    if (deviceTarget.flashingMethod === FlashingMethod.UART) {
+      return true;
+    }
+    if (
+      deviceTarget.flashingMethod === FlashingMethod.BetaflightPassthrough &&
+      device.platform === 'esp8285'
```
Comment (jurgelenas) [inline src/ui/views/ConfiguratorView/index.tsx:441]: ESP32 based receivers are not supported in this case?

File: src/ui/views/ConfiguratorView/index.tsx [hunk 2]
```diff
@@ -618,7 +648,14 @@ const ConfiguratorView: FunctionComponent<ConfiguratorViewProps> = (props) => {
   }, [buildInProgress, response]);
 
   const onBuild = () => sendJob(BuildJobType.Build);
-  const onBuildAndFlash = () => sendJob(BuildJobType.BuildAndFlash);
+  const onBuildAndFlash = () => {
+    if (forceFlash) {
+      sendJob(BuildJobType.ForceFlash);
```
Comment (jurgelenas) [inline src/ui/views/ConfiguratorView/index.tsx:653]: I think this is a good time to get rid of "ForceFlash" build job type. `force` and `erase` should be boolean flags. `BuildJobType` should retain `Build` and `Flash` only.

**Review by jurgelenas — CHANGES_REQUESTED** [review]

File: src/ui/views/ConfiguratorView/index.tsx (hunk 1 above)
Comment (pkendall64) [reply src/ui/views/ConfiguratorView/index.tsx:441]: No, because we are doing the flashing from our firmware rather than the boot rom.

**Comment by justinlampley** [comment]
@jurgelenas This is what it would look like if we move the checkboxes to the right of the Flash button.

image

My only issue is that it places options that will be used for an action after the button that initiates the action.


Another way we could handle these options is to remove the Build and Flash buttons and instead have users choose whether they want to Build or Flash as the first item in the Actions pane and a single button at the bottom that initiates the chosen action.  If they choose build, then we remove the options that are not applicable like selecting the Serial or WIFI Device, but if they choose flash, those are shown along with the other flashing options, Erase and Force Flash.

File: src/api/src/services/BinaryFlashingStrategy/BinaryConfigurator/index.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ export default class BinaryConfigurator {
     }
     flags.push(...this.userDefinesToFlags(params.userDefines));
 
-    if (params.type === BuildJobType.ForceFlash) {
```
Comment (jurgelenas) [inline src/api/src/services/BinaryFlashingStrategy/BinaryConfigurator/index.ts:54]: Hell yeah! Looks much better!

**Review by jurgelenas — APPROVED** [review]

**Pull request merged**